File: 01-array/0103-977.cpp
/*
 * leetcode 977
 *
 * double-pointer: left and right
 * the for-loop is not complete
 *
 */

class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int n = nums.size();
        int k = n-1;
        vector<int> res(n, 0);
        for (int i = 0, j = n-1; i <= j; ) {
            if (nums[i]*nums[i] < nums[j]*nums[j]) {
                res[k--] = nums[j]*nums[j];
                j--;
            } else {
                res[k--] = nums[i]*nums[i];
                i++;
            }
        }
        return res;
    }
};

class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int n = nums.size();
        int k = n-1;
        vector<int> res(n, 0);
        for (int i = 0, j = n-1; i <= j; ) {
            int x = nums[i]*nums[i];
            int y = nums[j]*nums[j];
            if (x < y) {
                res[k--] = y;
                j--;
            } else {
                res[k--] = x;
                i++;
            }
        }    
        return res;
    }
};

File: 01-array/0103-977.py
"""
leetcode 977
python

check it how the lambda function is used.

but this is brute-force solution.
it squares each element, then sort it.

"""

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        return sorted(list(map(lambda x: x*x, nums)))

class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        for i in range(len(nums)):
            nums[i] = nums[i]**2
        return sorted(nums)

# double-pointer
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        # double-pointer: left and right
        l, r = 0, len(nums)-1
        res = []
        while l <= r:
            x = nums[l]*nums[l]
            y = nums[r]*nums[r]
            if x >= y:
                res.append(x)
                l += 1
            else:
                res.append(y)
                r -= 1
        return reversed(res)

